File: Problems/518. Coin Change ll/memoization.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int solve(int amount, vector<int>& coins, int index, vector<vector<int>>& memo) {
        if (amount == 0) return 1;
        if (amount < 0 || index < 0) return 0;

        if (memo[amount][index] != -1) return memo[amount][index];

        int include = solve(amount - coins[index], coins, index, memo);
        int exclude = solve(amount, coins, index - 1, memo);

        return memo[amount][index] = include + exclude;
    }

    int change(int amount, vector<int>& coins) {
        int n = coins.size();
        vector<vector<int>> memo(amount + 1, vector<int>(n, -1));
        return solve(amount, coins, n - 1, memo);
    }
};


int main () {

    return 0;
}

File: Problems/62. Unique Paths/memoization.py
class Solution:
    def solve(self, m:int, n:int, i:int, j:int, memo) -> int:
        if i > m or j > n:
            return 0
        if i == m and j == n:
            return 1

        if (i, j) in memo:
            return memo[(i, j)]
            
        memo[(i,j)] = self.solve(m, n, i+1, j, memo) + self.solve(m, n, i, j+1, memo)
        return memo[(i,j)]

    def uniquePaths(self, m: int, n: int) -> int:
        memo = {}
        return self.solve(m-1, n-1, 0, 0, memo)